src/packages/configuration-service.ts

```diff
@@ -34,141 +36,162 @@ interface IReverseClientAssetMapping {
-                if (this.globalAssetMap[this.clientConfig.assetMapping[currencySymbol]]) {
-                    resolvedClientAssetMapping[currencySymbol] = this.globalAssetMap[this.clientConfig.assetMapping[currencySymbol]];
-                }
+    public getBlockstackServiceForConfig = async (userCruxID?: string, identityClaim?: IIdentityClaim): Promise<blockstackService.BlockstackService> => {
+        if (!(this.clientConfig && this.nameServiceConfig)) {
+            throw ErrorHelper.getPackageError(PackageErrorCode.ClientNotInitialized);
+        }
+        let nsConfig = this.nameServiceConfig;
+
+        // Merging user level configuration overrides if available (will be overriding the temporary in-memory variable)
+        if (userCruxID && identityClaim) {
+            // Get the user name registration status
+            const ns: blockstackService.BlockstackService = new blockstackService.BlockstackService(nsConfig);
+            await ns.restoreIdentity(userCruxID, identityClaim);
+            const status = await ns.getRegistrationStatus(identityClaim);
+
+            if (status.status === blockstackService.SubdomainRegistrationStatus.DONE) {
+                const userBlockstackID = identityUtils.IdTranslator.cruxToBlockstack(identityUtils.CruxId.fromString(userCruxID)).toString();
+                const gaiaUrls = await getGaiaDataFromBlockstackID(userBlockstackID, (this.clientConfig.nameserviceConfiguration && this.clientConfig.nameserviceConfiguration.bnsNodes) || config.BLOCKSTACK.BNS_NODES);
+                const gaiaHub = gaiaUrls.gaiaWriteUrl;
+                nsConfig = Object.assign(this.nameServiceConfig, {gaiaHub});
             }
         }
-        return resolvedClientAssetMapping;
+        return new blockstackService.BlockstackService(nsConfig);
+    }
+
+    public translateSymbolToAssetId = async (currencySymbol: string): Promise<string> => {
+        return (this.clientAssetMapping as IClientAssetMapping)[currencySymbol];
+    }
+
+    public translateAssetIdToSymbol = async (assetId: string): Promise<string> => {
+        return (this.reverseClientAssetMapping as IReverseClientAssetMapping)[assetId];
+    }
+
+    public getCruxIDByAddress = async (address: string) => {
+        if (!this.clientConfig) {
+            throw ErrorHelper.getPackageError(PackageErrorCode.ClientNotInitialized);
+        }
+        const walletNameServiceConfiguration = this.clientConfig.nameserviceConfiguration;
+        const bnsNodes = (walletNameServiceConfiguration && walletNameServiceConfiguration.bnsNodes) ? [...new Set([...config.BLOCKSTACK.BNS_NODES, ...walletNameServiceConfiguration.bnsNodes])] : config.BLOCKSTACK.BNS_NODES;
+        const registrar = (walletNameServiceConfiguration && walletNameServiceConfiguration.subdomainRegistrar) || config.BLOCKSTACK.SUBDOMAIN_REGISTRAR;
+        return await getCruxIDByAddress(this.clientName, address, bnsNodes, registrar);
     }
```

Comment:
Can you explain how are you visualizing ConfigService? 

Why such diverse functionality of getCruxIDByAddress, translateAssetIdToSymbol, translateSymbolToAssetId are inside same class?

Reply:
One way we can see is, ConfigService is an in-memory store for all configurations and business logic like above-mentioned functionality is outside.

getBlockstackServiceForConfig: This could be a static method inside BlockstackService which interacts with ConfigService to fetch all details and initializes its own object like a static constructor.

Reply:
seems like - 

All these functions cannot be executed without being aware of config logic - defaults and overrides etc.
Also ConfigService is the only entity aware of WalletName and hence has all the logic to do with the relationship between WalletName and CruxID. 

`getBlockstackServiceForConfig` -> This should be getNameServiceForCruxID.

It's more of a `WalletIntegrationManager` than a `ConfigService` I guess. 

`getCruxIDByAddress` makes sense because without knowing the Wallet name we can't `getCruxIDByAddress` because we need to know the registrar.